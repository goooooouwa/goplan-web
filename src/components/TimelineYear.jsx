import { Box, Container, Grid, Stack, Typography } from '@mui/material';
import moment from 'moment';
import React from 'react';
import TodoMonthSlider from './TodoMonthSlider';

const marks = [
  {
    value: 0,
    label: 'Jan',
  },
  {
    value: 1,
    label: 'Feb',
  },
  {
    value: 2,
    label: 'Mar',
  },
  {
    value: 3,
    label: 'Apr',
  },
  {
    value: 4,
    label: 'May',
  },
  {
    value: 5,
    label: 'Jun',
  },
  {
    value: 6,
    label: 'Jul',
  },
  {
    value: 7,
    label: 'Aug',
  },
  {
    value: 8,
    label: 'Sep',
  },
  {
    value: 9,
    label: 'Oct',
  },
  {
    value: 10,
    label: 'Nov',
  },
  {
    value: 11,
    label: 'Dec',
  },
];

export default function TimelineYear(props) {

  return (
    <>
    <Box sx={{ flexGrow: 1 }}>
      <Grid container spacing={2}>
        <Grid item xs={12} md={4}>
            <Typography
              variant="body1"
              gutterBottom
              sx={{
                fontWeight: 'bold',
                textAlign: 'left',
                mt: 1,
                ml: 2
              }}
            >
              Todos
            </Typography>
          </Grid>
          <Grid item xs={12} md={8}>
            <Box display={{ xs: 'none', md: 'block' }} m={1}>
              <Stack
                direction="row"
                justifyContent="space-between"
                alignItems="center"
              >
                {marks.map((mark, index) => (
                  <Typography
                    key={index}
                    variant="body1"
                    gutterBottom
                    sx={{
                      fontWeight: 'bold'
                    }}
                  >
                    {mark.label}
                  </Typography>
                ))}
              </Stack>
            </Box>
          </Grid>
          {props.todos
            .sort((t1, t2) => moment(t1.createdAt).isBefore(t2.createdAt) ? -1 : 1)
            .map((todo, index) => (
              <TodoMonthSlider key={index} todo={todo} />
            ))}
        </Grid>
      </Box>
    </>
  );
}
