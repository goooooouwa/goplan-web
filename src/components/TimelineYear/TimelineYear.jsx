import { Box, Grid, IconButton, Stack, Typography } from '@mui/material';
import isInYearRange from 'utils/rangeCheck';
import moment from 'moment';
import React from 'react';
import { Link as RouterLink, useParams } from 'react-router-dom';
import PropTypes from 'prop-types';
import momentPropTypes from 'react-moment-proptypes';
import SHARED_PROP_TYPES from 'utils/sharedPropTypes';
import TodoYearSlider from '../TodoYearSlider/TodoYearSlider';

const marks = [
  {
    value: 0,
    label: 'Jan',
  },
  {
    value: 1,
    label: 'Feb',
  },
  {
    value: 2,
    label: 'Mar',
  },
  {
    value: 3,
    label: 'Apr',
  },
  {
    value: 4,
    label: 'May',
  },
  {
    value: 5,
    label: 'Jun',
  },
  {
    value: 6,
    label: 'Jul',
  },
  {
    value: 7,
    label: 'Aug',
  },
  {
    value: 8,
    label: 'Sep',
  },
  {
    value: 9,
    label: 'Oct',
  },
  {
    value: 10,
    label: 'Nov',
  },
  {
    value: 11,
    label: 'Dec',
  },
];

export default function TimelineYear(props) {
  const params = useParams();
  const monthUrlPrefix = params.projectId !== undefined ? `/projects/${params.projectId}/month?month=` : '/timeline/month?month=';

  return (
    <>
      <Grid container rowSpacing={1} >
        <Grid item xs={12} md={4}>
          <Typography
            variant="body1"
            gutterBottom
            sx={{
              fontWeight: 'bold',
              textAlign: 'left',
              mt: 1,
              ml: 2
            }}
          >
            Todos
          </Typography>
        </Grid>
        <Grid item xs={12} md={8}>
          <Box display='block' m={1}>
            <Stack
              direction="row"
              justifyContent="space-between"
              alignItems="center"
            >
              {marks.map((mark, index) => (
                <IconButton
                  key={index}
                  component={RouterLink}
                  size="small"
                  to={`${monthUrlPrefix}${props.selectedYear.clone().add(mark.value, "months").format("YYYYMM")}`}
                  sx={{
                    fontWeight: 'bold',
                    fontSize: { xs: 10, sm: 16 },
                    color: (isInYearRange(moment(), props.selectedYear) && moment().month() === index) ? 'error.main' : 'text.primary'
                  }}
                >
                  {mark.label}
                </IconButton>
              ))}
            </Stack>
          </Box>
        </Grid>
        {props.todos
          .map((todo, index) => (
            <TodoYearSlider key={index} todo={todo} selectedYear={props.selectedYear} handleTodoChange={props.handleTodoChange} handleMonthChange={props.handleMonthChange} />
          ))}
      </Grid>
    </>
  );
}

TimelineYear.propTypes = {
  selectedYear: momentPropTypes.momentObj.isRequired,
  todos: PropTypes.arrayOf(SHARED_PROP_TYPES.todo).isRequired,
  handleTodoChange: PropTypes.func.isRequired,
  handleMonthChange: PropTypes.func.isRequired
};