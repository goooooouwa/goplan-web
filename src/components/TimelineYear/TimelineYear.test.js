import React from 'react';
import renderer from 'react-test-renderer';
import moment from 'moment';
import { BrowserRouter } from 'react-router-dom';
import TimelineYear from './TimelineYear';

const todos = [{
  id: 1,
  projectId: 1,
  status: false,
  repeat: true,
  name: "Todo 1",
  startDate: '2022-01-02',
  endDate: '2022-02-02',
  instanceTimeSpan: 1,
}];

it('should show a timeline Year view', () => {
  const component = renderer.create(
    <BrowserRouter>
      <TimelineYear todos={todos} selectedYear={moment('2022-02-01')} handleTodoChange={() => { }} handleMonthChange={() => { }} />
    </BrowserRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});