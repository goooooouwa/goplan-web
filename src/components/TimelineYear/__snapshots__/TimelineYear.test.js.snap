// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a timeline Year view 1`] = `
<div
  className="MuiGrid-root MuiGrid-container css-1sjizby-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lj5egr-MuiGrid-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-1xzmp4b-MuiTypography-root"
    >
      Todos
    </p>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-9drkqd-MuiGrid-root"
  >
    <div
      className="MuiBox-root css-4h9s5b"
    >
      <div
        className="css-3y1137-MuiStack-root"
      >
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202202"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Jan
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202203"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Feb
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202204"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Mar
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202205"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Apr
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202206"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          May
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-8lu1d-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202207"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Jun
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202208"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Jul
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202209"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Aug
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202210"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Sep
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202211"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Oct
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202212"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Nov
        </a>
        <a
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1cb383b-MuiButtonBase-root-MuiIconButton-root"
          href="/timeline/month?month=202301"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          Dec
        </a>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lj5egr-MuiGrid-root"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      disabled={false}
    >
      <span
        className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <input
          aria-label="controlled"
          checked={false}
          className="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate={false}
          onChange={[Function]}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
      <a
        className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/projects/1/todos/1"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <div
          className="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Todo 1
          </span>
        </div>
      </a>
    </li>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-1x9uoy6-MuiGrid-root"
  >
    <span
      className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
      onMouseDown={[Function]}
    >
      <span
        className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
      />
      <span
        className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
        style={
          Object {
            "left": "0%",
            "width": "9.090909090909092%",
          }
        }
      />
      <span
        className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
        data-index={0}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <span
        className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
        data-index={1}
        style={
          Object {
            "left": "9.090909090909092%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={2}
        style={
          Object {
            "left": "18.181818181818183%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={3}
        style={
          Object {
            "left": "27.272727272727273%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={4}
        style={
          Object {
            "left": "36.36363636363637%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={5}
        style={
          Object {
            "left": "45.45454545454545%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={6}
        style={
          Object {
            "left": "54.54545454545455%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={7}
        style={
          Object {
            "left": "63.63636363636363%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={8}
        style={
          Object {
            "left": "72.72727272727273%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={9}
        style={
          Object {
            "left": "81.81818181818181%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={10}
        style={
          Object {
            "left": "90.9090909090909%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={11}
        style={
          Object {
            "left": "100%",
          }
        }
      />
      <span
        className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
        data-index={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "0%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-orientation="horizontal"
          aria-valuemax={11}
          aria-valuemin={0}
          aria-valuenow={0}
          aria-valuetext={0}
          data-index={0}
          disabled={false}
          max={11}
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={0}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              Jan
            </span>
          </span>
        </span>
      </span>
      <span
        className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
        data-index={1}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "9.090909090909092%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-orientation="horizontal"
          aria-valuemax={11}
          aria-valuemin={0}
          aria-valuenow={1}
          aria-valuetext={1}
          data-index={1}
          disabled={false}
          max={11}
          min={0}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={1}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              Feb
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</div>
`;
